Add engine_from_config to sqlalchemy integration

pacejackson opened the pull request:
This allows us to specify the different connection params and fetch the password from a SecretsStore.

👓 @spladug @bsimpson63

pacejackson commented on baseplate/context/sqlalchemy.py:
```diff
@@ -3,14 +3,70 @@
-)
+from baseplate import config
+from baseplate.context import ContextFactory
+from baseplate.core import ServerSpan, SpanObserver
```
@spladug how do you feel about me cleaning up these imports while I'm already in here?

spladug replied:
i feel good. very good.

spladug commented on baseplate/context/sqlalchemy.py:
```diff
@@ -3,14 +3,70 @@
+    config_prefix = prefix[:-1]
+    cfg = config.parse_config(app_config, {
+        config_prefix: {
+            "drivername": config.String,
```
why not allow the url format?

pacejackson replied:
I kinda feel like that might get confusing about what you are supposed to set?  You can still use `engine_from_config` straight from sqlalchemy if you want which just uses the URL but then you have to embed the password in the URL.

pacejackson commented:
💇‍♂️ 
I changed it so you can either use `database.url` or `database.drivername`.  If you set the url, it just uses that, it does no go so far to see if you set anything else and raise an error if you do.

As a result, I also moved to make `secrets` an optional argument unless you actually set `password_secret`.

LMK if you think this makes sense or if it's too many conditional requirements (which was my initial concern).

spladug commented:
What about parsing the URL first then allowing any other config fields to override then kwargs, in that order. That seems like it'd make the most sense (since the common case would be url + password from secret).

spladug commented on baseplate/context/sqlalchemy.py:
```diff
@@ -3,14 +3,88 @@
+            "url": config.Optional(config.String),
+            "drivername": config.Optional(config.String),
+            "username": config.Optional(config.String),
+            "password_secret": config.Optional(config.String),
```
I think we probably want to have username be in here too. If the credentials are being rotated, that'd generally be done by changing user to a new one with a new password, rolling everyone off the old one, then killing the old user. Right?

spladug replied:
(do we want to add a "credentials" secret type?)

pacejackson replied:
🤔

pacejackson replied:
https://github.com/reddit/baseplate.py/pull/267

pacejackson commented:
> What about parsing the URL first then allowing any other config fields to override then kwargs, in that order. That seems like it'd make the most sense (since the common case would be url + password from secret).

That seems like it would get really confusing really quickly other than the case of overwriting the username/password?

pacejackson commented:
Maybe the thing to do is to _only_ support url + optional credentials? @spladug

pacejackson commented:
💇‍♂️ 
This simplifies it to only support `url` and `credentials` (note the tests will fail because I'm starting to use code from #267 ).

bsimpson63 commented:
🐟 this looks great

pacejackson commented:
💇‍♂️ 

![Screen Shot 2019-04-10 at 5 14 49 PM](https://user-images.githubusercontent.com/1929108/55922073-7499fb00-5bb4-11e9-9b17-de6b9b8e6183.png)
